### PR TITLE
SI-7039 if unapplySeq returns an option of a product `nbSubPats` is not ...

### PR DESCRIPTION
**Pull request by jrudolph**
...enough to check because this will miss a pattern with one product element and an empty repeated argument list.

The proposed solution is to check the number of elements in the expected product as well before concluding that an extractor is _no_ unapplySeq.

This seems to fix it but I suppose there are better ways to achieve the same effect.

### CODE REVIEW

**Comment by retronym** [comment]
Review by @adriaanm

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -88,7 +88,7 @@ trait Infer extends Checkable {
       else resTp.baseType(OptionClass).typeArgs match {
         case optionTArg :: Nil =>
           def productArgs = getProductArgs(optionTArg)
-          if (nbSubPats == 1) {
+          if ((nbSubPats == 1 && !isUnapplySeq) || productArgs.size == 1) {
             if (isUnapplySeq) List(seqToRepeatedChecked(optionTArg))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:92]: I don't understand this change. This means we'll call `seqToRepeatedChecked(optionTArg)` with `optionTArg =:= Product1[?T]`, but `seqToRepeatedChecked` is the identity for types that don't have `Seq` as a baseclass, and `Product1` doesn't inherit `Seq`.

I think the right solution is to make unapplySeq v unapply the top-level in the decision tree.

Closing this PR and submitting a fix along those lines.

**Comment by adriaanm** [comment]
Thanks for looking into this! You're right that this needs to be fixed, but I'd prefer to do it a little differently. PR coming.

**Comment by jrudolph** [comment]
Yes, of course, do it differently. I was happy I found a way to fix it at all but understood too little to make it right.

**Pull request closed without merging**